File: commitment/server/commitment_api/git_commands.ts
import { Command, logData, doNotLogData } from './command';

export const checkIfRepoExists = (url: string): Command => ({
  ...doNotLogData,
  cmd: `git ls-remote ${url}`, // tries to get a repo hash from a url
  onSuccess: (command: string) => `✅ Found Repo ${url}`,
  onFail: (command: string, error: Error) => `❌ Could not find Repo ${url}: ${error.message}`,
});

export const cloneRepo = (url: string, targetDirectory: string): Command => ({
  ...doNotLogData,
  cmd: `git clone ${url} \"${targetDirectory}\"`, // clone url to local path
  onSuccess: (command: string) => `✅ Repo successfully cloned to ${targetDirectory}`,
  onFail: (command: string, error: Error) => `❌ Error cloning repo: ${error.message}`,
});

export const getBranches = (): Command => ({
  ...doNotLogData,
  cmd: 'git branch -a', // gets only local branches
});

export const getAllCommitsFrom = (branch: string): Command => ({
  ...doNotLogData,
  cmd: `git log ${branch} --format=%H`, // gets all commits from this branch with the full 40 character SHA-1 hash for complete unambiguity
});

export const getContributorEmails = (name: string): Command => ({
  ...doNotLogData,
  cmd: `git log --author="${name}" --pretty=format:"%ae"`, // gets all contributor emails inside the repo
});

export const getCommitDetails = (hash: string) => ({
  ...doNotLogData,
  cmd: `git show --pretty=format:"%H%n %an%n %ad%n %s%n %b" --diff-filter=ADMRC --name-status ${hash}`,
});

export const getFileDataFromCommit = (hash: string, path: string): Command => ({
  ...doNotLogData,
  cmd: `git show ${hash}:${path}`,
});

export const getOldFileDataFromCommit = (hash: string, path: string): Command => ({
  ...doNotLogData,
  cmd: `git show ${hash}^:${path}`,
});

export const getRepoName = (): Command => ({
  ...doNotLogData,
  cmd: 'git remote get-url origin',
});

File: api/src/GitCommands.hs
--{-# LANGUAGE OverloadedStrings #-}

module GitCommands (
  quote, 
  delim,
  checkIfRepoExists,
  cloneRepo,
  checkIsGitDirectory,
  getBranches,
  getAllCommitsFrom,
  getContributorEmails,
  getCommitDetails,
  getCommitDiff,
  getRepoName
) where

import Command
import Control.Concurrent.STM ( TBQueue )

quote :: String -> String
quote s = "\"" ++ s ++ "\""

delim :: String
delim  = "\\n|||END|||"

checkIfRepoExists :: String -> Command
checkIfRepoExists url = doNotLogData
  { command = "git ls-remote " ++ quote url
  , onSuccess = \_ _ -> "Found Repo " ++ url
  , onFail = \_ e -> "Could not find Repo (" ++ url ++ "): " ++ e
  }

cloneRepo :: String -> FilePath -> Command
cloneRepo url targetDirectory = doNotLogData
  { command =
      "git -c credential.helper= -c core.askPass=true clone --bare "
      ++ quote url ++ " " ++ quote targetDirectory
  , env_vars = Just [("GIT_TERMINAL_PROMPT", "0")]
  , env_clean = ["GIT_ASKPASS","GCM_INTERACTIVE"]
  , onSuccess = \_ _ -> url ++ " successfully cloned to " ++ targetDirectory
  , onFail = \c e -> "Error cloning repo:\nCommand:\n" ++ c ++ "\nError Message:\n" ++ e
  }

checkIsGitDirectory :: FilePath -> Command
checkIsGitDirectory dir = doNotLogData
  { command = "git -C " ++ quote dir ++ " rev-parse --git-dir"
  }

getBranches :: Command
getBranches = doNotLogData
  { command = "git --no-pager branch -a --format=" ++ quote "%(refname:short)"
  }

getAllCommitsFrom :: String -> Command
getAllCommitsFrom branch = doNotLogData
  { command = "git --no-pager log " ++ quote branch ++ " --format=%H"
  }

getContributorEmails :: String -> Command
getContributorEmails name = doNotLogData
  { command = "git log \"--author=" ++ name ++ "\" \"--pretty=format:%ae\""
  }

getCommitDetails :: String -> Command
getCommitDetails hash = doNotLogData
  { command = "git show \"--pretty=format:%H" ++ delim ++ "%an" ++ delim ++ "%cI" ++ delim ++ "%s" ++ delim ++ "%b" ++ delim ++ "\" \"--name-status\" " ++ hash 
  }

getCommitDiff :: String -> Command
getCommitDiff hash = doNotLogData
  { command = "git --no-pager diff-tree -p " ++ hash 
  }

getRepoName :: Command
getRepoName = doNotLogData
  { command = "git remote get-url origin"
  }
